File: guests/java/dev/kaya/guests/Submit.java
package dev.kaya.guests;

import dev.kaya.KayaApp;
import dev.kaya.KayaGen;

/**
 * The submit scene from the JVM — guests/rust/submit.rs,
 * tools/scenes/submit.steps. Return in an entry, a search field and a
 * {@code submits} textarea publishes the field's text
 * (docs/submit-plan.md); a plain textarea's Return is its newline. The
 * app writes each submit into one label.
 */
public final class Submit {
    @KayaGen(key = "String")
    record SubmitThread(String title) {}

    public static void app() {
        KayaApp app = new KayaApp();

        app.build(tx -> {
            var threads = SubmitThreadKaya.collection(tx);
            KayaApp.Signal<String> sent = tx.signal("sent: -");

            tx.mount(tx.column(col -> {
                tx.label(sent).a11yId("sent"); // label#0
                KayaApp.Widget name =
                        tx.entry().placeholder("Name").a11yId("name"); // entry#0
                app.onSubmitted(name, (t, text) -> t.write(sent, "sent: " + text));
                KayaApp.Widget find =
                        tx.search().placeholder("Search").a11yId("find"); // search#0
                app.onSubmitted(find, (t, text) -> t.write(sent, "sent: " + text));
                KayaApp.Widget plain = tx.textarea().a11yId("plain"); // textarea#0
                app.onSubmitted(plain, (t, text) -> t.write(sent, "sent: " + text));
                KayaApp.Widget compose =
                        tx.textarea().submits().a11yId("compose"); // textarea#1
                app.onSubmitted(compose, (t, text) -> t.write(sent, "sent: " + text));
                for (var row : SubmitThreadKaya.rows(tx, threads)) {
                    row.label(row.title);
                    KayaApp.Node reply = row.entry();
                    row.setA11yId(reply, "reply");
                    app.onSubmitted(reply, (t, keys, text) ->
                            t.write(sent, "sent: " + keys.get(0) + ": " + text));
                }
            }));

            threads.insert(tx, "r1", new SubmitThread("First"));
            threads.insert(tx, "r2", new SubmitThread("Second"));
            return null;
        });

        app.dispatchLoop();
    }

    private Submit() {}
}

File: guests/go/submit/submit.go
// The submit scene (tools/scenes/submit.steps): Return in an entry, a
// search field and a Submits textarea publishes the field's text
// (docs/submit-plan.md S1); a plain textarea's Return is its newline.
// The app writes each submit into one label.
package submit

import (
	"fmt"

	kaya "dev.kaya/bindings/go"
)

//go:generate go run dev.kaya/cmd/kaya-gen -type Thread -key string
type Thread struct {
	Title string
}

func App() *kaya.App {
	app := kaya.NewApp()

	app.Build(func(tx *kaya.Tx) {
		sent := tx.Signal("sent: -")
		threads := ThreadCollection(tx)
		var reply kaya.Node

		tx.Mount(tx.Column(func() {
			tx.Label(sent).A11yID("sent")
			wrote := func(tx *kaya.Tx, text string) {
				tx.Write(sent, fmt.Sprintf("sent: %s", text))
			}
			tx.Entry(nil).Placeholder("Name").A11yID("name").OnSubmitted(wrote)
			tx.Search(nil).Placeholder("Search").A11yID("find").OnSubmitted(wrote)
			tx.Textarea(nil).A11yID("plain").OnSubmitted(wrote)
			tx.Textarea(nil).Submits().A11yID("compose").OnSubmitted(wrote)
			for row := range ThreadRows(tx, threads).All() {
				row.Label(row.Title())
				reply = row.Entry(nil)
				row.SetA11yID(reply, "reply")
			}
		}))

		reply.OnSubmitted(func(tx *kaya.Tx, keys []any, text string) {
			tx.Write(sent, fmt.Sprintf("sent: %v: %s", keys[0], text))
		})

		threads.Insert(tx, "r1", Thread{Title: "First"})
		threads.Insert(tx, "r2", Thread{Title: "Second"})
	})

	return app
}
